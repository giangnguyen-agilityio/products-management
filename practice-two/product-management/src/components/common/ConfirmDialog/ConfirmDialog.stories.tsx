import ProductContext from '@context/ProductContext/ProductContext'
import { MemoryRouter } from 'react-router-dom'
import { Story, Meta } from '@storybook/react'
import ConfirmDialog from './index'

export default {
  title: 'Practice Two/ConfirmDialog Component',
  component: ConfirmDialog,
  decorators: [
    Story => (
      <div
        style={{
          backgroundColor: '#ECB203',
          width: '100vw',
          height: '100%',
        }}
      >
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    closeConfirmDialog: { action: 'The close confirm dialog button clicked' },
  },
} as Meta<typeof ConfirmDialog>

const mockProductData = {
  listProduct: [
    {
      id: 'P01',
      name: 'iPhone 14 Pro Max',
      image: 'https://assets.nicepagecdn.com/d2cc3eaa/3159880/images/yttyy.jpg',
      discount: 11,
      oldPrice: 1263.8,
      newPrice: 1116.31,
      rate: 4.5,
      description:
        "The iPhone 14 Pro Max is Apple's flagship smartphone, featuring a Super Retina XDR display, A16 Bionic chip for powerful performance, and an advanced triple-camera system for exceptional photography. With premium design, Face ID, and MagSafe technology, it offers a top-tier user experience.",
    },
  ],
  isLoading: false,
  isLoadingMore: false,
  isReachingEnd: false,
  isError: false,
  addNewProduct: async () => {
    console.log('Add new product')
  },
  handleLoadMore: () => {
    console.log('Load more product')
  },
}

const Template: Story<typeof ConfirmDialog> = args => {
  return (
    <MemoryRouter>
      <ProductContext.Provider value={mockProductData}>
        <ConfirmDialog {...args} />
      </ProductContext.Provider>
    </MemoryRouter>
  )
}

export const TheConfirmDialog = Template.bind({})
TheConfirmDialog.args = {
  id: 'id_123',
  isConfirmDialogOpen: false,
}
