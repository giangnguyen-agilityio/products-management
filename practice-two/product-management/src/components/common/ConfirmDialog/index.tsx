// Libraries
import React, { useRef, useState } from 'react'
import {
  Text,
  Image,
  Center,
  Button,
  AlertDialog,
  AlertDialogBody,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogContent,
  AlertDialogOverlay,
  AlertDialogCloseButton,
} from '@chakra-ui/react'

// Assets
import deleteAction from '@assets/images/delete_Action.gif'

// Define the props interface for the ConfirmDialog component
interface ConfirmDialogProps {
  id: string
  isConfirmDialogOpen: boolean
  closeConfirmDialog: () => void
  onDelete: (id: string) => void
}

const ConfirmDialog: React.FC<ConfirmDialogProps> = ({
  id,
  isConfirmDialogOpen,
  closeConfirmDialog,
  onDelete,
}) => {
  const cancelRef = useRef<HTMLButtonElement | null>(null)
  const [disableButton, setDisableButton] = useState(false)

  // Function to handle delete
  const handleDelete = async () => {
    setDisableButton(true)
    await onDelete(id)
    setDisableButton(false)
    closeConfirmDialog()
  }

  // Render the ConfirmDialog component
  return (
    <AlertDialog
      isOpen={isConfirmDialogOpen}
      leastDestructiveRef={cancelRef}
      onClose={closeConfirmDialog}
      isCentered
      motionPreset="slideInBottom"
    >
      <AlertDialogOverlay />
      <AlertDialogContent background="background" color="primary">
        <AlertDialogHeader>Confirmation</AlertDialogHeader>
        <AlertDialogCloseButton
          className="close-dialog-button"
          border="small"
          color="primary"
          borderRadius="circle"
          _hover={{
            borderColor: 'primary',
          }}
          onClick={closeConfirmDialog}
        />
        <AlertDialogBody
          flexDirection="column"
          alignItems="center"
          textAlign="center"
        >
          {/* Display image and text to confirm deletion */}
          <Center>
            <Image
              loading="eager"
              src={deleteAction}
              alt="The image for delete action"
              width="200px"
              height="150px"
            />
          </Center>
          <Text marginTop={5} fontWeight="semibold" color="primary">
            Do you really want to delete this item?
            <br />
            This process cannot be undone.
          </Text>
        </AlertDialogBody>
        <AlertDialogFooter>
          {/* Cancel button */}
          <Button
            className="cancel-btn"
            aria-label="Cancel button"
            ref={cancelRef}
            onClick={closeConfirmDialog}
            color="textPrimary"
            variant="minimal"
          >
            Cancel
          </Button>
          {/* Delete button */}
          <Button
            className="delete-btn"
            aria-label="Delete button"
            onClick={handleDelete}
            marginLeft={3}
            isLoading={disableButton}
            variant="danger"
            _hover={{ opacity: 1 }}
          >
            Delete
          </Button>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialog>
  )
}

export default ConfirmDialog
