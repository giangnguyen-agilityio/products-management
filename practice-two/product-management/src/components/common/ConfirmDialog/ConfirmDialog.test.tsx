import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import ConfirmDialog from './'

describe('ConfirmDialog Component', () => {
  const mockOnDelete = jest.fn()
  const mockCloseConfirmDialog = jest.fn()

  jest.mock('@assets/images/delete_Action.gif', () => ({
    default: 'deleteAction',
  }))

  const defaultProps = {
    id: 'item_id',
    isConfirmDialogOpen: true,
    closeConfirmDialog: mockCloseConfirmDialog,
    onDelete: mockOnDelete,
  }

  it('renders correctly', () => {
    const { container } = render(<ConfirmDialog {...defaultProps} />)

    expect(screen.getByText('Confirmation')).toBeInTheDocument()
    expect(screen.getByLabelText('Cancel button')).toBeInTheDocument()
    expect(screen.getByLabelText('Delete button')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('calls onDelete when Delete button is clicked', async () => {
    render(<ConfirmDialog {...defaultProps} />)

    const deleteButton = screen.getByLabelText('Delete button')
    fireEvent.click(deleteButton)

    await waitFor(() => {
      expect(mockOnDelete).toHaveBeenCalledWith('item_id')
    })
  })

  it('closes the dialog when Cancel button is clicked', () => {
    render(<ConfirmDialog {...defaultProps} />)

    const cancelButton = screen.getByLabelText('Cancel button')
    fireEvent.click(cancelButton)

    expect(mockCloseConfirmDialog).toHaveBeenCalled()
  })

  it('disables the Delete button temporarily when clicked', async () => {
    render(<ConfirmDialog {...defaultProps} />)

    const deleteButton = screen.getByLabelText('Delete button')
    fireEvent.click(deleteButton)

    expect(deleteButton).toBeDisabled()

    await waitFor(() => {
      expect(deleteButton).not.toBeDisabled()
    })
  })
})
