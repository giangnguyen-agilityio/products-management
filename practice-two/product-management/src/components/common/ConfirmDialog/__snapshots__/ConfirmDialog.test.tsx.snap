// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog Component renders correctly 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
